client/core/exchangeratefetcher.go

```diff
@@ -212,7 +215,7 @@ func coinpapSlug(symbol, name string) string {
 	return strings.ToLower(strings.ReplaceAll(slug, " ", "-"))
 }
 
-func getInto(ctx context.Context, url string, thing interface{}) error {
+func getRates(ctx context.Context, url string, thing interface{}) error {
 	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
```

Comment:
This function is where we need to check the response code.  Messari will be returning 429 instead of 200, but we don't seem to catch that at all.

Reply:
It's just rate limiting.  If you are sharing a public IP, you will get limited faster.
What can we do?  Just return an error and prevent some invalid value from being used.  We can still have the `<=0` check you added, but we know right here if it's valid or not.
I'm not keen to spam the logs though.

Reply:
Darn, I think we are hitting the rate limit. Assuming we make one request each for all 7 assets every 5min, we end up making ~2016 (7 * 12 req per hour * 24hr) requests in 24hrs although we have a 1000 daily limit.

And then this also affects `Coinparika` because for a user with 7 assets, we end up making `~60480` requests per month which exceed `Coinparika` monthly limit of `25_000` requests.

Reply:
I think we should just increase the request interval a bit, and handle non-200-status code responses properly.  Also cases where all sources fail to return a rate.

This affects all the sources, not just Messari.  As I mentioned, if you are sharing a public IP (e.g. on a VPN or Tor where hundreds or thousands of other users have the same source IP), you'll have failed API requests frequently.

We can also consider backing off in the individual rate fetchers if 429 codes are received.  For example, if `fetchDcrdataRates` gets a 429, it could toggle some global atomic flag assigned to the fetcher to skip future requests until the flag is cleared, and when setting that flag start a time.AfterFunc to clear it after say 5 minutes.

Reply:
> We can also consider backing off in the individual rate fetchers if 429 codes are received. For example, if fetchDcrdataRates gets a 429, it could toggle some global atomic flag assigned to that fetcher to skip future requests until the flag is cleared, and when setting that flag start a time.AfterFunc to clear it after say 5 minutes.

Yeah, this will make more sense if the rate limiting is based on a minute-by-minute throttling but then, we are never hitting any of the per-minute limits(at least not until we add support for up to 20 assets).

Reply:
Depends if the service keeps counting failed requests against your rate limit.  Anyway, just a thought.  Let's go with what you have now.